### ISSUE TITLE
Flagging LHS data quality

### ISSUE THREAD

**Issue by mwinokan** [m2ms/fragalysis-frontend#1408]
We need to allow data curators to upload preliminary experimental observations and have this be flagged in the f/e. Maybe a new tag group is sufficient?

**Comment by mwinokan** [comment]
@mwinokan to come up with a name for the new tag category

**Comment by mwinokan** [comment]
@Waztom says that tagging is a sensible first step. But how do we communicate more granular density-related issues for the ligand. I.e. labelling which atoms/bonds are particularly problematic

**Comment by mwinokan** [comment]
@phraenquex says the atom-level annotation cannot be part of the green release.

**Comment by mwinokan** [comment]
@phraenquex says that as a basic quick win:

- Traffic lights for each observation (green, orange, red)
- How do we manage clashes between different curators? I.e. might be green for one person, and orange for another
- Do we show the history? Overview of who tagged the observation with which score?

**Comment by mwinokan** [comment]
@matej-vavrek, @boriskovar-m2ms and I mocked this up. @phraenquex I hope it fits the bill

Summary:

- Traffic light indicating tie breaker / overall status in the LHS hit navigator
- If you hover over that light you get a summary of the "peer review"
- If you click on the light or the ellipsis you will get a bigger dialog where you can set the overall status and see the history of peer review changes with comments
- If you click the plus you can add your own review with a comment

![Image](https://github.com/user-attachments/assets/bfff9aef-a70d-4b19-a42f-9ce0e4f9d2c9)

**Comment by mwinokan** [comment]
@mwinokan to fill out spec:

- Ticket for LHS work @matej-vavrek 
- Ticket for schema changes @kaliif (endpoints `siteobservation` and whatever the new model is) 
- Ticket for b/e changes for the auto-build scores

**Comment by mwinokan** [comment]
@kaliif has been working on the backend changes for this, but please review compatibility with the latest mockup/spec above

**Comment by mwinokan** [comment]
Thanks @kaliif 

If you want to use a checkbox/boolean for auto-assigned then that's fine. As long as @matej-vavrek says that doesn't cause any issues for the f/e.

Regarding the difference between the status change and peer review that is indeed subtle/confusing. @phraenquex was quite clear that there should be a way to provide a privileged opinion/status over a peer review. Maybe you could only have one model but include a Boolean field like "main_status" or something like that

**Comment by kaliif** [comment]
@mwinokan is this `main_status` always expected to be the most recent one? Or is there a valid case where a non-privileged opinion comes later?

**Comment by phraenquex** [comment]
Correct, `main_status` is a separate thing.  Certainly not the latest one, it can only be set by a privileged user.

@mwinokan I forget what we discussed as the F/E actions to set one or the other - but a peer opinion shoulnd't automatically change `main_status`.

**Comment by mwinokan** [comment]
@phraenquex agreed. A peer review should not change the main status. 

However, we currently do not maintain any concept of privileged users so for now any authenticated user on the proposal could set the main_status.

**Comment by phraenquex** [comment]
Sure.  But setting `main_status` should be a separate action to simply adding a peer-review.  

(Do rn this past Daren, and maybe all of SBCore at tomorrow's meeting.)

**Comment by mwinokan** [comment]
@phraenquex in the mockup changing the main status and adding a peer review are quite separate actions

**Comment by kaliif** [comment]
Ok @matej-vavrek, here's the first iteration of the backend: `kaliif/fragalysis-backend:m2ms-1408-lhs-quality`. When you're working with it, load some new data; the loader (randomly) assignes statuses to observations, so you'll have some data to play with.

New models:
- QualityStatusType
- SiteObservationQualityStatus

QualityStatusType is automatically populated with values
- GOOD
- MEDIOCRE
- BAD
- NONE

As you can see I went with adding flags for auto-assigned state and main observation, let me know if that's a problem or if anything else is missing.

Endpoint `/api/site_observation_quality` allows read-write access to the model. There's no patch/update at the moment. @mwinokan I have some vague recollection that comment field was supposed to be editable, is that correct or was that about something else? If yes, I'll add update option for comment field.

**Comment by mwinokan** [comment]
Amazing @kaliif, thank you. Please do add the update functionality for the comment fields, I'm sure that will be useful

**Comment by matej-vavrek** [comment]
@mwinokan, @phraenquex, do these statuses need to be "live"? i.e. if a target will be opened by two people and they will add a status separately if it should check for updates and load them (like current service status feed)

**Comment by mwinokan** [comment]
@matej-vavrek ideally, yes. If it is a massive development overhead let us know please

**Comment by mwinokan** [comment]
@kaliif please adjust the endpoint to allow for an editable comment field

**Comment by kaliif** [comment]
@mwinokan just a quick clarification, should the comments be updatable only by users who created them?

**Comment by mwinokan** [comment]
@kaliif yes I think so

**Comment by mwinokan** [comment]
@boriskovar-m2ms reports that @matej-vavrek has started working on this

**Comment by mwinokan** [comment]
@matej-vavrek has started to work on it and a preview is coming soon! (~tomorrow) This will include some backend communication but will not be feature complete.

@matej-vavrek thinks that all the b/e features required have already be implemented so all the work is frontend.

**Comment by matej-vavrek** [comment]
@mwinokan, I am sorry but I had to switch to another project so this will be a bit delayed

**Comment by Waztom** [comment]
@matej-vavrek will get back on Thursday. If you need any input from Frank, need to deploy today/soonest. There is a deadline for ILO meeting Wed 26th Feb. Need time to iterate. Matej will try to push to Tibor's stack.

**Comment by matej-vavrek** [comment]
@mwinokan, @phraenquex, I have updated [my stack](https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/).

Two things came to my mind:
- comment can not be blank (main status does not have any, I am putting "-" for it at the moment)
- user is only a number or ..? @kaliif

**Comment by mwinokan** [comment]
Excellent @matej-vavrek, thanks.

Some suggestions from me:

- The selected set status light should be the current status:

<img width="291" alt="Image" src="https://github.com/user-attachments/assets/36ecf65d-a14f-4f90-b1fd-ab215b07b441" />

- There should be a way to know which observation's status you are setting. Maybe just a title bar for the modal with `A0237a Peer Review` or make it behave more like the observations modal which is pinned to the row being edited.
- There are some incorrect pluralisations in the times, e.g. `1 minutes ago`
- Could you have the status change in the table of peer review instead, like in the mockup? I.e. when user 1 changed their mind from green to red it should be a green light with a little arrow to red. @phraenquex please chime in if this is too confusing then.

<img width="275" alt="Image" src="https://github.com/user-attachments/assets/cf0361b7-38d1-4b11-8028-e70383a0c433" />

Thanks!

**Comment by phraenquex** [comment]
Thanks @matej-vavrek.  Three things (see image):
1. The detailed modal should stay right up against the overview modal
2. that check-mark looks like a "okay-close-this" button.  Will need fixing - in fact, it's not at all self-explanatory how it's meant to work.  Shouldn't it be greyed-out (or invisible) until one of the statuses next to it is checked?
3. The status light feels too small - can it max out the width of that box?

![Image](https://github.com/user-attachments/assets/4035b957-a7d7-4e8e-8cef-8900858be78d)

**Comment by phraenquex** [comment]
Setting the status should add a line to the peer-review list, with auto-comment.

**Comment by matej-vavrek** [comment]
> Setting the status should add a line to the peer-review list, with auto-comment.

@phraenquex should it also be part of peer review "statistics" or add it only in the list for having a status history?

**Comment by phraenquex** [comment]
Add it to the statistics.

**Comment by mwinokan** [comment]
Summarising @matej-vavrek and @mwinokan discussions on Slack:

- main status changes should be part of the peer review table but visually distinct (bold/italic/color)
- Main status changes should not be counted in the peer review statistics/summary lights

**Comment by kaliif** [comment]
>     * user is only a number or ..? [@kaliif](https://github.com/kaliif)

@matej-vavrek what do you need? Username or full name?

**Comment by mwinokan** [comment]
@matej-vavrek Thanks! looks much better.

One extra thing on the pluralisation: `0 minutes` should be plural

![Image](https://github.com/user-attachments/assets/7d3d8004-e4de-44c5-8185-7485fdce28bd)

**Comment by phraenquex** [comment]
@matej-vavrek looks great!

Two more things
* I shouldn't be able to make any changes unless logged in.
* Very important to see _who_ made each change (in the details modal).

Both are blockers, I think.

**Comment by matej-vavrek** [comment]
@mwinokan, thank you, I have fixed it and also pizzas were added.
It goes clockwise through individual peer review statuses in order GOOD, MEDIOCRE, BAD, NONE.
I am thinking about changing the yellow colour for pizza to some with more contrast, what do you think?

![Image](https://github.com/user-attachments/assets/f723099b-74b9-4239-815d-4322398ca8a9)

> what do you need? Username or full name?

@kaliif, I am sorry, I missed this comment.
Will it be still created by user id or will it be changed to the chosen one like we get what we send?
Because I can only see username besides user id in the django context but I am not very familiar with a "user" management here.

**Comment by phraenquex** [comment]
@kaliif please provide both Username and full name.
@matej-vavrek show the full name, and put the username into a tooltip.

**Comment by mwinokan** [comment]
Thanks @matej-vavrek. We'll review the pizza styling and colours in a meeting later today and get back to you

**Comment by mwinokan** [comment]
Tagging as deliverable as this is needed for deposition workflow

**Comment by mwinokan** [comment]
Thanks @kaliif for updating the endpoint to provide user names and fed ID's. @matej-vavrek will work on this later today.

@matej-vavrek please also work on restricting the peer review to logged in / authenticated users only

**Comment by matej-vavrek** [comment]
@mwinokan, I have added user names as Frank described to [my stack](https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/).
Is it enough to just remove possibility to add main status and peer review for not logged in user? (this is also already there)

**Comment by mwinokan** [comment]
Thanks @matej-vavrek 

Yes please lock the adding/editing of main status/peer review behind the same login/authentication requirements as editing/adding tags

**Comment by matej-vavrek** [comment]
@mwinokan, @phraenquex, is there something more to adjust or did I miss some point or is this alright for now?

**Comment by mwinokan** [comment]
@matej-vavrek there are two issues that appeared today:

- The data that was visible when not logged in, disappeared after logging in
- The user names are showing as `undefined`:

![Image](https://github.com/user-attachments/assets/68c0fdd1-ea0c-41f3-b8f2-d9074b631cdd)

If these two issues are resolved we can go to staging to test more thoroughly

**Comment by matej-vavrek** [comment]
@mwinokan, it is undefined because it does not know at the moment a name of the user.
We can either force reload data from `api/site_observation_quality` or @kaliif can add fields for the response (preferred for me)

![Image](https://github.com/user-attachments/assets/2886fac7-bb66-4037-bcf1-57e84c504433)

![Image](https://github.com/user-attachments/assets/53003c13-8eb2-4515-acce-3873685818fb)

**Comment by mwinokan** [comment]
@matej-vavrek would this be an issue if we started with a fresh target? I.e. is it just a problem because some of the peer review / statuses were created before the endpoints were updated?

**Comment by matej-vavrek** [comment]
@mwinokan, no. The problem is that at the moment when the peer review / statuses are created the F/E does not have the info about the user (only ID or username from DJANGO_CONTEXT).

**Comment by phraenquex** [comment]
So the action is with @kaliif ?  (Three comments up.)

@kaliif, for your priorities:  this ticket is a blocker.

**Comment by matej-vavrek** [comment]
@kaliif, just if the POST response of `api/site_observation_quality` (after creating) could return same object as GET, i.e. add `username`, `first_name` and `last_name` to it, please

**Comment by mwinokan** [comment]
@matej-vavrek what is the status of this ticket, please?

**Comment by matej-vavrek** [comment]
@mwinokan, I would say it will be done after the B/E update for the mentioned POST response of `api/site_observation_quality`

**Comment by mwinokan** [comment]
@kaliif please make the changes to the endpoint that @matej-vavrek requested: https://github.com/m2ms/fragalysis-frontend/issues/1408#issuecomment-2703128832

**Comment by kaliif** [comment]
@matej-vavrek aren't  names are already in the request? Seems to be the case in staging. The ones that are null are auto-created on load so there wouldn't be a name.

**Comment by matej-vavrek** [comment]
@mwinokan, I have redeployed [my stack](https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/) with latest B/E changes. Could you check it now, please?

**Comment by mwinokan** [comment]
Thanks @matej-vavrek

![Image](https://github.com/user-attachments/assets/afb9b4a8-cbc1-49e4-a3d6-50b7b98990cf)

Where's the best place to fix the "Created on load" behaviour? It should not be shown at all. Is it better to fix this in the backend or frontend?

**Comment by matej-vavrek** [comment]
@mwinokan I think we do not need to create them. They were populated just to have some data or @kaliif ?

**Comment by mwinokan** [comment]
@phraenquex advises to handle this in the f/e. @matej-vavrek please ignore the loader-generated "Created on load" blank status changes

@phraenquex is unsure if the Fed-ID name mapping is confidential, but if there is no reason to have them they should be removed. @matej-vavrek please remove the tooltip

**Comment by matej-vavrek** [comment]
@mwinokan I have updated [my stack](https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/) and after that I also added this information about no peer reviews (this is not there yet)

![Image](https://github.com/user-attachments/assets/3b43a8ce-0283-4a09-9edf-f93055e40cbc)

**Comment by phraenquex** [comment]
@matej-vavrek great!!

Final _tiny_ change:  move the three-dots to the left-side of the preview dialog - that so that I don't have to move the mouse, but can simply click, to bring up the detailed peer-review modal.

![Image](https://github.com/user-attachments/assets/18b56766-8056-4440-a2af-5f799434af11)

**Comment by mwinokan** [comment]
@matej-vavrek hopefully not to confuse the spec, but I think it would be more elegant as @phraenquex says that if you click the traffic light it opens up the modal. So I think the ellipsis (three dots) can stay where they are, and the hover/tooltip just needs to go slightly to the right so it doesn't overlap with the original traffic light.

**Comment by mwinokan** [comment]
@matej-vavrek is back on fragalysis and will be able to deploy to his stack later today

**Comment by matej-vavrek** [comment]
@mwinokan, I have updated [my stack](https://fragalysis-matej-default.xchem-dev.diamond.ac.uk/) with the latest change. As I am looking at it right now, maybe we can also add a tooltip for the main light - "Click to view peer reviews" or something like that?

Edit: I have updated my stack with the tooltip, but the behaviour is a bit weird (maybe because all the hover actions over there) and it does not seem very user friendly to me..

**Comment by mwinokan** [comment]
Thanks @matej-vavrek the changes look great. But now for some reason I can't see which light I have selected when trying to change the main status:

<img width="637" alt="Image" src="https://github.com/user-attachments/assets/40951f65-86f4-45f8-a30c-992e502ec586" />

If you can fix that, then it's ready for staging

**Comment by matej-vavrek** [comment]
@mwinokan thank you for noticing, I have fixed that and should the tooltip be like this or better none or just a classic one which disappears?

![Image](https://github.com/user-attachments/assets/09523ab8-387f-47cb-922f-f2aed52f9efd)

**Comment by phraenquex** [comment]
@matej-vavrek have you updated your stack?  Because now it's very non-obvious when I've clicked a new status, or that the tick is now active.  Maybe put a border around the clicked status, and the tick?  And make it much fainter when it's not activated, and darker when it's activated.

**Comment by matej-vavrek** [comment]
@phraenquex, it is what Max mentioned. I have not redeployed it but I can in a moment

Edit: @phraenquex, @mwinokan, it should be fixed on my stack now

**Comment by mwinokan** [comment]
Thanks @matej-vavrek. 

Hopefully a final thing, the hover-summary stays open even when you click out of the modal, and doesn't disappear until you mouse over the status light. Please make it disappear together with the modal so you don't end up with something like this:

![Image](https://github.com/user-attachments/assets/84da80b8-3569-4e9b-890f-765fa6dda7b7)

**Comment by mwinokan** [comment]
@matej-vavrek has redeployed but found another small bug regarding the pop-up not always showing due to overlapping elements
